### PR TITLE
Simplify memory ordering intrinsics

### PR DESCRIPTION
**Pull request by m-ou-se**
This changes the names of the atomic intrinsics to always fully include their memory ordering arguments.

```diff
- atomic_cxchg
+ atomic_cxchg_seqcst_seqcst

- atomic_cxchg_acqrel
+ atomic_cxchg_acqrel_release

- atomic_cxchg_acqrel_failrelaxed
+ atomic_cxchg_acqrel_relaxed

// And so on.
```

- `seqcst` is no longer implied
- The failure ordering on chxchg is no longer implied in some cases, but now always explicitly part of the name.
- `release` is no longer shortened to just `rel`. That was especially confusing, since `relaxed` also starts with `rel`.
- `acquire` is no longer shortened to just `acq`, such that the names now all match the `std::sync::atomic::Ordering` variants exactly.
- This now allows for more combinations on the compare exchange operations, such as `atomic_cxchg_acquire_release`, which is necessary for #68464.
- This PR only exposes the new possibilities through unstable intrinsics, but not yet through the stable API. That's for [a separate PR](https://github.com/rust-lang/rust/pull/98383) that requires an FCP.

Suffixes for operations with a single memory order:

| Order   | Before       | After      |
|---------|--------------|------------|
| Relaxed | `_relaxed`   | `_relaxed` |
| Acquire | `_acq`       | `_acquire` |
| Release | `_rel`       | `_release` |
| AcqRel  | `_acqrel`    | `_acqrel`  |
| SeqCst  | (none)       | `_seqcst`  |

Suffixes for compare-and-exchange operations with two memory orderings:

| Success | Failure | Before                   | After              |
|---------|---------|--------------------------|--------------------|
| Relaxed | Relaxed | `_relaxed`               | `_relaxed_relaxed` |
| Relaxed | Acquire | :x:                      | `_relaxed_acquire` |
| Relaxed | SeqCst  | :x:                      | `_relaxed_seqcst`  |
| Acquire | Relaxed | `_acq_failrelaxed`       | `_acquire_relaxed` |
| Acquire | Acquire | `_acq`                   | `_acquire_acquire` |
| Acquire | SeqCst  | :x:                      | `_acquire_seqcst`  |
| Release | Relaxed | `_rel`                   | `_release_relaxed` |
| Release | Acquire | :x:                      | `_release_acquire` |
| Release | SeqCst  | :x:                      | `_release_seqcst`  |
| AcqRel  | Relaxed | `_acqrel_failrelaxed`    | `_acqrel_relaxed`  |
| AcqRel  | Acquire | `_acqrel`                | `_acqrel_acquire`  |
| AcqRel  | SeqCst  | :x:                      | `_acqrel_seqcst`   |
| SeqCst  | Relaxed | `_failrelaxed`           | `_seqcst_relaxed`  |
| SeqCst  | Acquire | `_failacq`               | `_seqcst_acquire`  |
| SeqCst  | SeqCst  | (none)                   | `_seqcst_seqcst`   |

### CODE REVIEW

**Comment by m-ou-se** [comment]
cc @tmiasko

**Comment by m-ou-se** [comment]
This updates the stdarch submodule to include https://github.com/rust-lang/stdarch/commit/dc4b2e25df41e1b6507cc156e89b09fab70dbc63, which isn't merged in stdarch's main branch yet, since that change needs to happen at the same time as the rest of this PR.

File: library/core/src/intrinsics.rs [hunk 1]
```diff
@@ -78,142 +286,226 @@ extern "rust-intrinsic" {
...
+    /// The stabilized version of this intrinsic is available on the
+    /// [`atomic`] types via the `compare_exchange` method by passing
+    /// [`Ordering::Acqrel`] and [`Ordering::Acquire`] as the success and failure parameters.
+    /// For example, [`AtomicBool::compare_exchange`].
+    pub fn atomic_cxchg_acqrel_acquire<T: Copy>(dst: *mut T, old: T, src: T) -> (T, bool);
+    /// Stores a value if the current value is the same as the `old` value.
+    ///
+    /// The stabilized version of this intrinsic is available on the
+    /// [`atomic`] types via the `compare_exchange` method by passing
+    /// [`Ordering::Acqrel`] and [`Ordering::SeqCst`] as the success and failure parameters.
```
Comment (tmiasko) [inline library/core/src/intrinsics.rs:366]: s/Acqrel/AcqRel/ here and in other places.

**Review by tmiasko — APPROVED** [review]

**Comment by tmiasko** [comment]
Looks good to me. Thanks! 

r=me with commits squashed.

**Comment by m-ou-se** [comment]
> This updates the stdarch submodule to include [rust-lang/stdarch@dc4b2e2](https://github.com/rust-lang/stdarch/commit/dc4b2e25df41e1b6507cc156e89b09fab70dbc63), which isn't merged in stdarch's main branch yet, since that change needs to happen at the same time as the rest of this PR.

cc @Amanieu for `stdarch` ^

**Comment by tmiasko** [comment]
> This updates the stdarch submodule to include [rust-lang/stdarch@dc4b2e2](https://github.com/rust-lang/stdarch/commit/dc4b2e25df41e1b6507cc156e89b09fab70dbc63), which isn't merged in stdarch's main branch yet, since that change needs to happen at the same time as the rest of this PR.

Since the `core_arch` uses intrinsics declared in the core, for the `not(bootstrap)` case we can also add re-exports under old name for those handful of intrinsics that are still required, and update stdarch using the usual process at some later point.

**Comment by matthiaskrgr** [comment]
looks like this failed in a rollup; https://github.com/rust-lang/rust/pull/98532#issuecomment-1166514086
@bors r-

**Comment by kkysen** [comment]
It seems that the error messages for these updated intrinsics are out-of-date with the new names, as they say ``similarly named function `{old_instrinsic_name}` defined here`` while pointing to the definition of the new intrinsic name.  For example,

```shell
error[E0425]: cannot find function `atomic_xchg_acq` in module `core::intrinsics`
   --> src/atomics.rs:167:33
    |
167 |         ) = ::core::intrinsics::atomic_xchg_acq(&mut x, 33 as libc::c_int);
    |                                 ^^^^^^^^^^^^^^^ help: a function with a similar name exists: `atomic_cxchg_acq`
    |
   ::: /home/kkysen/.rustup/toolchains/nightly-2022-08-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/intrinsics.rs:320:5
    |
320 |     pub fn atomic_cxchg_acquire_acquire<T: Copy>(dst: *mut T, old: T, src: T) -> (T, bool);
    |     -------------------------------------------------------------------------------------- similarly named function `atomic_cxchg_acq` defined here

error[E0425]: cannot find function `atomic_store_rel` in module `core::intrinsics`
   --> src/atomics.rs:171:25
    |
171 |     ::core::intrinsics::atomic_store_rel(&mut x, 0);
    |                         ^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `atomic_store`
    |
   ::: /home/kkysen/.rustup/toolchains/nightly-2022-08-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/intrinsics.rs:523:5
    |
523 |     pub fn atomic_store_seqcst<T: Copy>(dst: *mut T, val: T);
    |     -------------------------------------------------------- similarly named function `atomic_store` defined here
```

**Pull request merged**